tiledb/sm/filesystem/s3.cc

```diff
@@ -1066,6 +1165,171 @@ Status S3::write(const URI& uri, const void* buffer, uint64_t length) {
+  auto& state = state_iter->second;
+
+  auto& intermediate_chunks = state.buffered_chunks;
+  uint64_t sum_sizes = 0;
+  std::vector<uint64_t> offsets;
+  for (auto& chunk : intermediate_chunks) {
+    offsets.push_back(sum_sizes);
+    sum_sizes += chunk.size;
+  }
+
+  // Account for the current chunk too
+  offsets.push_back(sum_sizes);
+  sum_sizes += length;
+
+  if (sum_sizes < tiledb::sm::constants::s3_min_multipart_part_size) {
+    auto new_uri = generate_chunk_uri(uri, intermediate_chunks.size());
+    throw_if_not_ok(write_direct(new_uri, buffer, length));
+    intermediate_chunks.emplace_back(new_uri.to_string(), length);
+    return;
+  }
+
+  // We have enough buffered for a multipart part upload
+  std::vector<std::byte> merged(sum_sizes);
+  uint64_t length_returned;
+  auto st = parallel_for(
+      vfs_thread_pool_, 0, intermediate_chunks.size(), [&](size_t i) {
+        throw_if_not_ok(read(
+            URI(intermediate_chunks[i].uri),
+            0,
+            merged.data() + offsets[i],
+            intermediate_chunks[i].size,
+            0,
+            &length_returned));
+        return Status::Ok();
+      });
+  std::memcpy(merged.data() + offsets.back(), buffer, length);
+
+  // Issue one more multiple part uploads depending on the cumulative size of
+  // the merged chunks. It's a valid case that one of the submit() calls brings
+  // here a 10GB attribute write request, so we need to account for it and issue
+  // parallel writes if necessary
+  uint64_t num_ops = utils::math::ceil(
+      merged.size(),
+      std::max(
+          multipart_part_size_,
+          tiledb::sm::constants::s3_min_multipart_part_size));
+  num_ops = std::min(std::max(num_ops, uint64_t(1)), max_parallel_ops_);
+
+  auto upload_id = state.upload_id;
+
+  // Assign the part number(s), and make the write request.
+  if (num_ops == 1) {
+    const int part_num = state.part_number++;
+    auto ctx = make_upload_part_req(
+        aws_uri, merged.data(), merged.size(), upload_id, part_num);
+    throw_if_not_ok(get_make_upload_part_req(uri, uri_path, ctx));
+  } else {
+    std::vector<MakeUploadPartCtx> ctx_vec;
+    ctx_vec.resize(num_ops);
+    const uint64_t bytes_per_op = multipart_part_size_;
+    const int part_num_base = state.part_number;
+    auto st = parallel_for(vfs_thread_pool_, 0, num_ops, [&](uint64_t i) {
+      uint64_t begin = i * bytes_per_op,
+               end = std::min(
+                   (i + 1) * bytes_per_op - 1, uint64_t(merged.size() - 1));
+      uint64_t thread_nbytes = end - begin + 1;
+      auto thread_buffer = reinterpret_cast<const char*>(merged.data()) + begin;
+      int part_num = static_cast<int>(part_num_base + i);
+      ctx_vec[i] = make_upload_part_req(
+          aws_uri, thread_buffer, thread_nbytes, upload_id, part_num);
+
+      return Status::Ok();
+    });
+    state.part_number += num_ops;
+
+    for (auto& ctx : ctx_vec) {
+      auto st = get_make_upload_part_req(uri, uri_path, ctx);
```

Comment:
NIT: we could get rid of st and do if (!get_make_upload_part_req(uri, uri_path, ctx).ok()) {